File: .claude/hooks/pytest.mjs
#!/usr/bin/env node
// @hookstack stop-pytest
// Exécute pytest à la fin d'une session Python (Stop)
import { existsSync } from 'fs';
import { spawnSync } from 'child_process';
import { fileURLToPath } from 'url';

const PYTHON_MARKERS = ['pyproject.toml', 'setup.py', 'pytest.ini', 'setup.cfg'];

export function run({
  exists = existsSync,
  spawn = spawnSync,
  cwd = process.env.CLAUDE_PROJECT_DIR ?? process.cwd(),
} = {}) {
  const isPython = PYTHON_MARKERS.some(f => exists(`${cwd}/${f}`));
  if (!isPython) return null;

  const hasXdist = spawn('uv', ['run', 'python', '-c', 'import xdist'], {
    encoding: 'utf8',
    timeout: 10_000,
    cwd,
    stdio: 'ignore',
  }).status === 0;

  const pytestArgs = hasXdist
    ? ['run', 'pytest', '-n', 'auto', '--tb=short', '-q']
    : ['run', 'pytest', '--tb=short', '-q'];

  const result = spawn('uv', pytestArgs, {
    encoding: 'utf8',
    timeout: 300_000,
    cwd,
    stdio: ['ignore', 'pipe', 'pipe'],
  });

  const out = (result.stdout ?? '') + (result.stderr ?? '');
  const status = result.status ?? 1;
  const message = status !== 0
    ? `[pytest] ÉCHEC (exit ${status})\n${out.slice(-2000)}\n`
    : `[pytest] ✓ Tests passés\n${out.split('\n').filter(Boolean).slice(-5).join('\n')}\n`;

  return { status, message };
}

/* v8 ignore next 6 */
if (process.argv[1] === fileURLToPath(import.meta.url)) {
  const result = run();
  if (result) {
    process.stderr.write(result.message);
    if (result.status !== 0) process.exit(2);
  }
}

File: .claude/hooks/pytest.py
#!/usr/bin/env python3
# @hookstack stop-pytest
"""Runs pytest at the end of a Python session (Stop)."""
import os
import re
import subprocess
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
from lib.changed_files import changed_files  # noqa: E402

PYTHON_MARKERS = ["pyproject.toml", "setup.py", "pytest.ini", "setup.cfg"]
PY = re.compile(r"\.py$")
PY_CFG = re.compile(r"(^|/)(pyproject\.toml|pytest\.ini|setup\.cfg|setup\.py)$")


def _spawn(args, *, timeout=300, cwd=None):
    result = subprocess.run(
        ["uv", *args],
        capture_output=True,
        text=True,
        timeout=timeout,
        cwd=cwd,
    )
    return result


def run(*, exists=None, spawn=None, cwd=None, changed=None):
    if cwd is None:
        cwd = os.environ.get("CLAUDE_PROJECT_DIR") or os.getcwd()
    if exists is None:
        exists = lambda f: os.path.exists(os.path.join(cwd, f))
    if changed is None:
        changed = changed_files(cwd=cwd)
    if spawn is None:
        spawn = _spawn

    # CI (GitHub Action générée par le CLI) : lancer toute la suite, le scope
    # « fichiers modifiés » n'a pas de sens sur un arbre git propre.
    if os.environ.get("HOOKSTACK_FULL_CHECK") == "1":
        changed = None

    is_python = any(exists(f) for f in PYTHON_MARKERS)
    if not is_python:
        return None

    # Aucun .py (ni config pytest) modifié → inutile de relancer toute la suite.
    if changed is not None and not any(PY.search(f) or PY_CFG.search(f) for f in changed):
        return None

    has_xdist = (
        spawn(
            ["run", "python", "-c", "import xdist"],
            timeout=10,
            cwd=cwd,
        ).status
        == 0
    )

    pytest_args = (
        ["run", "pytest", "-n", "auto", "--tb=short", "-q"]
        if has_xdist
        else ["run", "pytest", "--tb=short", "-q"]
    )

    result = spawn(pytest_args, timeout=300, cwd=cwd)
    out = (result.stdout or "") + (result.stderr or "")
    status = getattr(result, "status", None)
    if status is None:
        status = result.returncode if hasattr(result, "returncode") else 1
    message = (
        f"[pytest] ÉCHEC (exit {status})\n{out[-2000:]}\n"
        if status != 0
        else f"[pytest] ✓ Tests passés\n{out.splitlines()[-5:]}\n"
    )

    return {"status": status, "message": message}


if __name__ == "__main__":
    result = run()
    if result:
        sys.stderr.write(result["message"])
        if result["status"] != 0:
            sys.exit(2)
